### PR TITLE
Shard and replia crds initial commit

### PR DESCRIPTION
**Pull request by sepulworld**

### CODE REVIEW

**Comment by sepulworld** [comment]
Initial start for SolrShard. Will follow up in this PR for replica CRD

```
apiVersion: solr.bloomberg.com/v1beta1
kind: SolrReplica
metadata:
  name: example-extra-replica-to-shard
  namespace: default
spec:
  solrCloud: example
  collection: example-collection-3-implicit
  shard: example-extra-shard
  node: "default-example-solrcloud-1.ing.local.domain:80_solr"
```

will be the idea

File: api/v1beta1/solrshard_types.go [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+	// Node set that the shard is for
+	// +optional
+	NodeSet []string `json:"nodeSet,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// SolrShard is the Schema for the solrshards API
+type SolrShard struct {
+	metav1.TypeMeta   `json:",inline"`
```
Comment (kcmartin) [inline api/v1beta1/solrshard_types.go:53]: Extra comma?

File: api/v1beta1/solrshard_types.go [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+
+	Spec   SolrShardSpec   `json:"spec,omitempty"`
+	Status SolrShardStatus `json:"status,omitempty"`
+}
+
+// +kubebuilder:object:root=true
+
+// SolrShardList contains a list of SolrShard
+type SolrShardList struct {
+	metav1.TypeMeta `json:",inline"`
```
Comment (kcmartin) [inline api/v1beta1/solrshard_types.go:64]: Extra comma?

File: controllers/suite_test.go [hunk 3]
```diff
@@ -94,6 +94,9 @@ var _ = BeforeSuite(func(done Done) {
 	err = solrv1beta1.AddToScheme(scheme.Scheme)
 	Expect(err).NotTo(HaveOccurred())
 
+	err = solrv1beta1.AddToScheme(scheme.Scheme)
+	Expect(err).NotTo(HaveOccurred())
```
Comment (kcmartin) [inline controllers/suite_test.go:98]: Isn't this a duplicate of L94-95 ?

File: controllers/util/collection_util.go [hunk 4]
```diff
@@ -85,6 +85,59 @@ func CreateCollecttionAlias(cloud string, alias string, aliasType string, collec
...
+	queryParams := url.Values{}
+	NodeSetArray := strings.Join(createNodeSet, ",")
+	queryParams.Add("action", "CREATESHARD")
+	queryParams.Add("shard", shard)
+	queryParams.Add("collection", collection)
+	queryParams.Add("createNodeSet", NodeSetArray)
+
+	resp := &SolrAsyncResponse{}
+
+	log.Info("Calling to create shard", "namespace", namespace, "cloud", cloud, "shard", shard, "collections", collection, "createNodeSet", createNodeSet)
```
Comment (kcmartin) [inline controllers/util/collection_util.go:99]: "collections" plural, should be singular?

File: controllers/util/collection_util.go [hunk 5]
```diff
@@ -85,6 +85,59 @@ func CreateCollecttionAlias(cloud string, alias string, aliasType string, collec
...
+	err = CallCollectionsApi(cloud, namespace, queryParams, resp)
+
+	if err == nil {
+		if resp.ResponseHeader.Status == 0 {
+			log.Info("ResponseHeader.Status", "ResponseHeader.Status", resp.ResponseHeader.Status)
+			success = true
+		}
+	} else {
+		log.Error(err, "Error creating shard", "namespace", namespace, "cloud", cloud, "shard", shard, "collections", collection, "createNodeSet", createNodeSet)
```
Comment (kcmartin) [inline controllers/util/collection_util.go:108]: "collections" plural, should be singular?

**Review by kcmartin — CHANGES_REQUESTED** [review]

File: api/v1beta1/solrshard_types.go (hunk 1 above)
Comment (sepulworld) [reply api/v1beta1/solrshard_types.go:53]: comma is expected here

File: controllers/suite_test.go (hunk 3 above)
Comment (sepulworld) [reply controllers/suite_test.go:98]: Yeah, good find, bunch of duplicates, ill clean up!

File: controllers/util/collection_util.go (hunk 4 above)
Comment (sepulworld) [reply controllers/util/collection_util.go:99]: yep, good looking out!

**Pull request closed without merging**